### PR TITLE
Build, publish using CircleCI

### PR DESCRIPTION
**Pull request by abesto**
_openzipkin/zipkin on CircleCI_: https://circleci.com/gh/openzipkin/zipkin

Why: https://github.com/openzipkin/zipkin/issues/1237
## Points of interest
- Support for the test-runtime-based balancing of tests CircleCI does for running tests in parallel
- Caching the `node_modules` dir of `zipkin-ui` saves around 90 seconds
- Overall build time is around 6-7 minutes (about half of that on Travis)
- Release process drafted in #1250, still needs thinking.

### CODE REVIEW

**Comment by abesto** [comment]
CC @adriancole @shakuzen @marcingrzejszczak @eirslett

File: circleci/publish-stable.sh [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+#     This commit will not trigger a separate build due to the [skip ci] in the
+#     commit message.
+./mvnw --batch-mode -s ./.settings.xml -Prelease -nsu -pl -:benchmarks \
+       -DskipTests -DscmCommentPrefix='[maven-release-plugin][skip ci] ' \
+       deploy
+
+# And this syncs it to Maven Central. Note: this needs to be done once per
+# project, not module, hence -N
+./mvnw --batch-mode -s ./.settings.xml -nsu -N \
+       io.zipkin.centralsync-maven-plugin:centralsync-maven-plugin:sync
```
Comment (abesto) [inline circleci/publish-stable.sh]: AFAICT on Travis, this runs on the commit with the non-SNAPSHOT version. But here by the time we get to this point, we're on version (release + 0.0.1)-SNAPSHOT. I suspect we'll need to expose a `sync.version` property so we can pass in the version to sync. @adriancole WDYT?

**Comment by codefromthecrypt** [comment]
Thanks for putting this together. I've a head-start on this as toyed a
little last couple days. Here's one thought..

separate out the generic changes that make tests not flake into a commit.
separate out the non-publish changes into a separate commit (ex ones that
make circle work)
add a new PR for the publishing things.. (or separate ^^ into a different
PR and keep this one alive)

Thought process is that we can merge having dual pull request checkers
(existing one and now circleci). We could then carry on a separate
discussion around the publish oriented things, including the more crude
credential tooling etc.

We didn't find travis to be a problem until time passed. My overall concern
is quickly moving without experience in CircleCI, we end up worse than
travis. One way we could address this is have pull request checkers here,
but use a less critical repo for guinea-pig testing release (ex
zipkin-dependencies, which doesn't change that often)

Thoughts?

**Comment by marcingrzejszczak** [comment]
That's a good idea. In Spring Cloud we've basically only done simple builds in Travis so our transition could have been extreme (create a CircleCI build and disable Travis ASAP). Here the process can be more gradual definitely.

**Comment by abesto** [comment]
I like this approach. Will separate out publishing into a separate PR, and squash this into two commits (one the CircleCI setup, the other changes to Zipkin code for the Circle env).

Side question: the flakes that we had on Circle, are those things our users would experience in their (non-test) environments?

**Comment by codefromthecrypt** [comment]
> Side question: the flakes that we had on Circle, are those things our
> users would experience in their (non-test) environments?
> 
> The flake fixes for things that happen in travis and circle might be OS
> specific or resource specific. OSX laptop builds don't fail for me often.

**Pull request merged**